dex/networks/dcr/script.go

```diff
@@ -120,216 +120,62 @@ const (
-// pay-to-alt-pubkey-hash script.
-func IsPubKeyHashAltScript(script []byte) bool {
-	pk, _ := ExtractPubKeyHashAltDetails(script)
-	return pk != nil
-}
-
-// ExtractPubKeyHash extracts the pubkey hash from the passed script if it is a
-// /standard pay-to-pubkey-hash script.  It will return nil otherwise.
-func ExtractPubKeyHash(script []byte) []byte {
-	// A pay-to-pubkey-hash script is of the form:
-	//  OP_DUP OP_HASH160 <20-byte hash> OP_EQUALVERIFY OP_CHECKSIG
-	if len(script) == 25 &&
-		script[0] == txscript.OP_DUP &&
-		script[1] == txscript.OP_HASH160 &&
-		script[2] == txscript.OP_DATA_20 &&
-		script[23] == txscript.OP_EQUALVERIFY &&
-		script[24] == txscript.OP_CHECKSIG {
-
-		return script[3:23]
-	}
-
-	return nil
-}
-
-// ExtractScriptHash extracts the script hash from the passed script if it is a
-// standard pay-to-script-hash script. It will return nil otherwise. See
-// ExtractStakeScriptHash for stake transaction p2sh outputs.
-//
-// NOTE: This function is only valid for version 0 opcodes.  Since the function
-// does not accept a script version, the results are undefined for other script
-// versions.
-func ExtractScriptHash(script []byte) []byte {
-	// A pay-to-script-hash script is of the form:
-	//  OP_HASH160 <20-byte scripthash> OP_EQUAL
-	if len(script) == 23 &&
-		script[0] == txscript.OP_HASH160 &&
-		script[1] == txscript.OP_DATA_20 &&
-		script[22] == txscript.OP_EQUAL {
-
-		return script[2:22]
-	}
-	return nil
-}
-
-// ExtractPubKeyHashAltDetails extracts the public key hash and signature type
-// from the passed script if it is a standard pay-to-alt-pubkey-hash script.  It
-// will return nil otherwise.
-func ExtractPubKeyHashAltDetails(script []byte) ([]byte, dcrec.SignatureType) {
-	// A pay-to-alt-pubkey-hash script is of the form:
-	//  DUP HASH160 <20-byte hash> EQUALVERIFY SIGTYPE CHECKSIG
-	//
-	// The only two currently supported alternative signature types are ed25519
-	// and schnorr + secp256k1 (with a compressed pubkey).
-	//
-	//  DUP HASH160 <20-byte hash> EQUALVERIFY <1-byte ed25519 sigtype> CHECKSIG
-	//  DUP HASH160 <20-byte hash> EQUALVERIFY <1-byte schnorr+secp sigtype> CHECKSIG
-	//
-	//  Notice that OP_0 is not specified since signature type 0 disabled.
-
-	if len(script) == 26 &&
-		script[0] == txscript.OP_DUP &&
-		script[1] == txscript.OP_HASH160 &&
-		script[2] == txscript.OP_DATA_20 &&
-		script[23] == txscript.OP_EQUALVERIFY &&
-		isStandardAltSignatureType(script[24]) &&
-		script[25] == txscript.OP_CHECKSIGALT {
-
-		return script[3:23], dcrec.SignatureType(asSmallInt(script[24]))
-	}
-
-	return nil, 0
-}
-
-// Check if the opcode is one of a small number of acceptable stake opcodes that
-// are prepended to P2PKH scripts. Return the opcode if it is, else OP_0.
-func stakeOpcode(script []byte) byte {
-	if len(script) == 0 {
-		return 0
-	}
-	opcode := script[0]
-	if opcode == txscript.OP_SSGEN || opcode == txscript.OP_SSRTX {
```

Comment:
Ahh, the `IsStakePubkeyHashScript` and `IsStakeScriptHashScript` were not detecting ticket outputs as stake and lumping them as nonstandard and thus not spendable.
We need to allow `InputInfo` to only return a non-nil `*SpendInfo` for the actually spendable stake output types.

Reply:
It previously reported vote outputs, so it is strange that we didn't have trouble because those have maturity rules that were never checked.

Reply:
Well, I'm now checking the `ListUnspentResult.Spendable` field in the funding and spendable utxo parsing methods in https://github.com/decred/dcrdex/pull/1308/commits/4bc82f77293d79da2126fc0cd3cc8c0b619fb7d5
We do the same in BTC.
I'm actually really surprised that immature and thus unspendable `stakegen` outputs from votes never threw these errors.

Reply:
> We need to allow InputInfo to only return a non-nil *SpendInfo for the actually spendable stake output types.

The above is still on the table as a complement to the `ListUnspentResult.Spendable` in the calling code, but I've not done this.  It would mean calling various stake output types "unsupported" script classes even though we can identify them as stake, so it feels icky going back to that behavior.  Might yet though...